### PR TITLE
Factor out Khiops and task argument preprocessing in a separate function

### LINKED ISSUE #208: detect_format is not ignored when headerline and separator are provided
#### Description
In many API functions, the detect_format flag is available. It defaults to true, so users providing headerline and separator are sometimes unaware that is is set.
**Problem**: when both detect_format and headerline + separator are set, the detection is run even if its results are ignored. For some datatables, the detection produces confusing warnings which causes users to wonder if the processing was done with the right separator.
**Example output**:
warning : "/usr/local/lib/python3.10/dist-packages/runner/khiops_runner.py:465: UserWarning: Khiops ended correctly but there were minor issues: Warnings in log:
Line 2: warning : Database format detector s3://bkt-guporionis-backend-dev-ow-idh-fe/orionis/deployment/caller/2024/06/06/0800-1200/flows/*.csv : 1 other field separator is possible: '-'
  runner.run(args)".

#### Questions/Ideas
- Looking at the code it seems that outputs from the detection are ignored but detection is run nevertheless - so detect_format flag is not ignored 
- If this behavior is desired, the doc should be updated and user instructed to ignore such messages
- Else (preferred) fix the behavior of the flag: if both headerline and separator are set, detect format should be forced to false
- Also, the doc is misleading about the mixed case: if detect_format is true and only headerline is set, it will detect both and headerline set by the user will override the detection, allowing to perform partial detection

### PR DESCRIPTION
**Pull request by popescu-v**
closes #208

### CODE REVIEW

File: khiops/core/api.py [hunk 1]
```diff
@@ -107,6 +107,41 @@ def _get_or_create_execution_dictionary_file(dictionary_file_path_or_domain, tra
     return execution_dictionary_file_path
 
 
+def preprocess_arguments(args):
```
Comment (folmos-at-orange) [inline khiops/core/api.py:110]: For better locality move after `_run_task` and before `_preprocess_task_arguments`.

File: khiops/core/api.py [hunk 2]
```diff
@@ -107,6 +107,41 @@ def _get_or_create_execution_dictionary_file(dictionary_file_path_or_domain, tra
...
+    """Preprocessing of Khiops arguments
+
+    Parameters
+    ----------
+    args : dict
+        The Khiops arguments.
+
+    Returns
+    -------
+    (`CommandLineOptions` instance, bool) tuple, with:
```
Comment (folmos-at-orange) [inline khiops/core/api.py:120]: The pattern in this type of documentation is 
```
tuple
    A 2-tuple containing:
    - A `~.CommandLineOptions` instance
    - A `bool` that is ``True`` whether the `dictionary_file_or_domain` is a `~.DictionaryDomain` instance
```

See for example `train_predictor`.

File: khiops/core/api.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:110]: Also, since this refactoring is only to acomodate `khrun` (which will be replaced) I'd mark the function as private (`_` prefix).

File: khiops/core/api.py [hunk 3]
```diff
@@ -107,6 +107,41 @@ def _get_or_create_execution_dictionary_file(dictionary_file_path_or_domain, tra
...
+        output_scenario_path=(
+            args["output_scenario_path"] if "output_scenario_path" in args else ""
+        ),
+        task_file_path=(args["task_file_path"] if "task_file_path" in args else ""),
+    )
+
+    # Clean the args to leave only the task arguments
+    _clean_task_args(args)
+
+    return (command_line_options, task_is_called_with_domain)
```
Comment (folmos-at-orange) [inline khiops/core/api.py:142]: The parentheses are not necessary.

**Review by folmos-at-orange — CHANGES_REQUESTED** [review]
See above.

File: khiops/core/api.py (hunk 2 above)
Comment (popescu-v) [reply khiops/core/api.py:120]: Yes, right, my bad. I would fix this.

File: khiops/core/api.py (hunk 3 above)
Comment (popescu-v) [reply khiops/core/api.py:142]: Indeed; I would remove them.

File: khiops/core/api.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/api.py:110]: > For better locality move after `_run_task` and before `_preprocess_task_arguments`.

Ok.

Comment (popescu-v) [reply khiops/core/api.py:110]: > Also, since this refactoring is only to acomodate `khrun` (which will be replaced) I'd mark the function as private (`_` prefix). We will probably defactor this in the future.

Yeah, we could still use it while it is necessary, then de-factor it, indeed. However, having it public and properly documented would help with code understanding a(nd reuse as well), IMHO.

Comment (folmos-at-orange) [reply khiops/core/api.py:110]: This will be used only once in code that is not public. It is refactored only for convenience but is not part of the API (even the "internals" API).

File: khiops/core/api.py (hunk 2 above)
Comment (popescu-v) [reply khiops/core/api.py:120]: Done.

File: khiops/core/api.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/api.py:110]: Done for the first point. Not convinced for the second point (see my comment above).

File: khiops/core/api.py (hunk 3 above)
Comment (popescu-v) [reply khiops/core/api.py:142]: Done.

File: khiops/core/api.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/api.py]: Done & Done (both points).

**Review by folmos-at-orange — APPROVED** [review]

**Comment by popescu-v** [comment]
Pushed alpha tag 10.2.2.5-a.1 on the feature branch of this PR.

**Comment by tramora** [comment]
Thank you for the change.

I used this pre-release as a base for modifications in a POC version of `khiops-api-base/kh-run`

Indeed eventually it fixed the customer issue : https://github.com/KhiopsML/khiops/issues/345 because it now produces these directives 
```
TrainDatabase.HeaderLineUsed true
TrainDatabase.FieldSeparator
```
instead of the previous ones 
```
TrainDatabase.HeaderLineUsed true
TrainDatabase.FieldSeparator 
TrainDatabase.DatabaseFormatDetector.DetectFileFormat
```
So we do not face any longer the issue khiops core must fix in the next release.

A few remarks though we may have discussed already :

## Public function to use

Both `khiops.core.api._preprocess_arguments` and `khiops.core.api._preprocess_tasks_arguments` have the private _ prefix and thus by convention are not officially exposed. 

In `khiops-api-base/kh-run`  the second `khiops.core.api._preprocess_tasks_arguments` is to be used in a loop of the `api_calls`.

## Asserts & default values

Because of asserts in code, the caller has the burden to set the default values of the 2 parameters `header_line` and `field_separator` along with `detect_format` before calling the API. It would have been nice if the API could handle the "business rules" internally.

## Semantic

The initial input protobuf payload contains only the `detect_format=True` param. It is quite surprising to end with a `HeaderLineUsed true` in the generated scenario even if the docstring states it clearly (I was forced to set the default values of `header_line` and `field_separator`)

```
It's ignored if ``header_line`` or ``field_separator`` are set.
```

**Comment by tramora** [comment]
As the produced output was incorrect (see the previous comment about the semantic), I tweaked in the last commit the `khiops-api-base/kh-run` code in order to both keep the semantic and to try to avoid ambiguity for khiops.

However the ambiguities remain (in **bold** below) and we are waiting for a khiops release to fix the issue https://github.com/KhiopsML/khiops/issues/345 about the `DetectFileFormat` 

| Input to `khiops-api-base/kh-run`| Output to khiops after processing by the API | Notes |
|---|---|---|
|*no input*|defect_format=True, header_line=True, field_separator=''|**Room for improvement here (for human readers) : set header_line to False as detect_format will take precedence**|
| header_line=True|defect_format=False, header_line=True, field_separator=''|
|field_separator=','|defect_format=False, header_line=True, field_separator=','|**Room for improvement here, header_line should be left False as it does not correlate with the modified field**|
| detect_format=True|defect_format=True, header_line=True, field_separator=''|**Room for improvement here (for human readers) : set header_line to False as detect_format will take precedence**|
| detect_format=True, field_separator=',' |defect_format=False, header_line=True, field_separator=','|**Room for improvement here, header_line should be left False as it does not correlate with the modified field**|

**Comment by popescu-v** [comment]
> Thank you for the change.
> 
> I used this pre-release as a base for modifications in a POC version of `khiops-api-base/kh-run`
> 
> Indeed eventually it fixed the customer issue : [KhiopsML/khiops#345](https://github.com/KhiopsML/khiops/issues/345) because it now produces these directives
> 
> ```
> TrainDatabase.HeaderLineUsed true
> TrainDatabase.FieldSeparator
> ```
> 
> instead of the previous ones
> 
> ```
> TrainDatabase.HeaderLineUsed true
> TrainDatabase.FieldSeparator 
> TrainDatabase.DatabaseFormatDetector.DetectFileFormat
> ```
> 
> So we do not face any longer the issue khiops core must fix in the next release.
> 
> A few remarks though we may have discussed already :
> ## Public function to use
> 
> Both `khiops.core.api._preprocess_arguments` and `khiops.core.api._preprocess_task_arguments` have the private _ prefix and thus by convention are not officially exposed.
> 
> In `khiops-api-base/kh-run` the second `khiops.core.api._preprocess_task_arguments` is to be used in a loop of the `api_calls`.
> 
> Branch : https://gitlab.tech.orange/khiops/khiops-api-base/-/tree/preprocess_task_arguments?ref_type=heads
> ## Asserts & default values
> 
> Because of asserts in code, the caller has the burden to set the default values of the 2 parameters `header_line` and `field_separator` along with `detect_format` before calling the API. It would have been nice if the API could handle the "business rules" internally.
> ## Semantic
> 
> The initial input protobuf payload contains only the `detect_format=True` param. It is quite surprising to end with a `HeaderLineUsed true` in the generated scenario even if the docstring states it clearly (I was forced to set the default values of `header_line` and `field_separator`).
> 
> ```
> (detect_format) ignored if ``header_line`` or ``field_separator`` are set.
> ```

Yes, here it should be ```is set to ``False`` if `header_line`` or ``field_separator`` are set```. On the other hand, if `header_line None`, then `header_line` is set to `True` in all cases; this where the documentation seems inexact, as it states that this should be the case only if `detect_format` is set to `False`.

File: khiops/core/api.py [hunk 4]
```diff
@@ -328,11 +328,11 @@ def _preprocess_format_spec(detect_format, header_line, field_separator):
...
-        - Disables ``detect_format`` if either ``header_line`` or ```field_separator``
-          are set
-        - If either ``header_line`` or ``field_separator`` is None then they are set to
-          their default values
-        - It transforms the field separator "\t" to the empty string ""
+        - Sets ``detect_format`` to ``False`` if either ``header_line`` or
+          ``field_separator`` are set
+        - If either ``header_line`` or ``field_separator`` is ``None``, then they are
+          set to their default values
+        - It transforms the field separator "\\t" to the empty string ""
```
Comment (folmos-at-orange) [inline khiops/core/api.py:335]: The second `\` is not necessary: we are in a raw string `r"..."`

File: khiops/core/api.py [hunk 5]
```diff
@@ -527,14 +527,15 @@ def build_dictionary_from_data_table(
...
-        as "\\t").
+        its field separator. It is set to ``False`` if ``header_line`` or
+        ``field_separator`` are set.
+    header_line : bool, optional (default ``None``)
+        It is set to ``True`` if ``None``.
+        If ``True`` it uses the first line of the data as column names. Sets
+        ``detect_format`` to ``False`` if set.
+    field_separator : str, optional (default ``None``)
+        A field separator character. It is set to "" if ``None``. "" has the same
+        effect as "\\t". Sets ``detect_format`` to ``False`` if set.
```
Comment (folmos-at-orange) [inline khiops/core/api.py:538]: I'd keep the old phrasing of the default value ("default `True` if ..."). This way the user is immediately informed about the default value and doesn't have to look below for the side effect.

I'm ok with the "Sets `detect_format` .. to ``False``" change.

File: khiops/core/api.py
Comment (folmos-at-orange) [inline khiops/core/api.py:1]: General comment on the doc commit: Why the message says that in _now_ reflects the current behavior? It is just a rephrasing.

File: khiops/core/api.py (hunk 4 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:335]: And all below.

File: .github/workflows/api-docs.yml [hunk 6]
```diff
@@ -9,7 +9,7 @@ on:
         type: boolean
         default: false
       image-tag:
-        default: latest
+        default: 10.2.3-b.5.0
```
Comment (folmos-at-orange) [inline .github/workflows/api-docs.yml:12]: General Comment: I'd rephrase the commit
Set _in CI_ Khiops core packages version to the latest beta

**Review by folmos-at-orange — CHANGES_REQUESTED** [review]

File: khiops/core/api.py
Comment (popescu-v) [reply khiops/core/api.py:1]: IMHO, no: before it was stated that:
- `detect_format` is ignored if at least one of `header_line` or `field_separator` is set; whereas, in reality, it set to `False`, which has the consequence of not performing automatic format detection on the binary Khiops side.
- `header_line` has `True` as default value, which is not strictly true, because the default value is `None`, in which case the attribute is set to `True` *afterwards*; the current discourse seems misleading, in that it creates the expectation that the default value is `True`;
- similar situation for `field_separator`, where the default value is not "\t", but `None`, in which case the attribute is set to "" *afterwards*; and "\t" and "" are equivalent *for the Khiops binaries* only.

So, the goal of this commit is to make sure the documentation reflects current behavior and semantics.

File: .github/workflows/api-docs.yml (hunk 6 above)
Comment (popescu-v) [reply .github/workflows/api-docs.yml:12]: Yes, OK for the addition, thus the intent is clearer.

File: khiops/core/api.py (hunk 5 above)
Comment (popescu-v) [reply khiops/core/api.py:538]: IMHO, it is better that the documentation reflects *exactly* the behavior (w/o shortcuts); otherwise, it seems misleading (see the other related comment). So, I would keep `None` as default value, with the statement of the side effect: longer-winded, but more correct IMHO. And IMHO this becomes even more important for client usage.

As for breaking the pattern for `header_line`: IMHO we are obliged to do so, in order to establish the "link" between the default `None` and the consequent default `True` as a side-effect. Otherwise, the documentation is not faithful to the implemented behavior IMHO.

File: khiops/core/api.py (hunk 4 above)
Comment (popescu-v) [reply khiops/core/api.py:335]: Ok, I will recheck in all cases and change where mandated by the `r"..."`.

Comment (popescu-v) [reply khiops/core/api.py:335]: Done.

File: .github/workflows/api-docs.yml (hunk 6 above)
Comment (popescu-v) [reply .github/workflows/api-docs.yml:12]: Done.

File: khiops/core/api.py (hunk 5 above)
Comment (popescu-v) [reply khiops/core/api.py]: Done as discussed.

File: khiops/core/api.py
Comment (popescu-v) [reply khiops/core/api.py:1]: Removed the "reflects the current behavior" part.

File: khiops/core/api.py [hunk 7]
```diff
@@ -510,14 +527,16 @@ def build_dictionary_from_data_table(
...
-    field_separator : str, optional (default "\\t" if ``detect_format`` is ``False``)
-        A field separator character, overrides ``detect_format`` if set ("" counts
-        as "\\t").
+        its field separator. It is set to ``False`` if ``header_line`` or
+        ``field_separator`` are set.
+    header_line : bool, optional (default ``True``)
+        If ``True`` it uses the first line of the data as column names. Sets
+        ``detect_format`` to ``False`` if set. Ignored if ``detect_format``
+        is set to ``True``.
+    field_separator : str, optional (default "")
```
Comment (folmos-at-orange) [inline khiops/core/api.py:536]: Set the default value to `\t` as before. Again, this _is_ its semantic value. And all below.

File: khiops/core/api.py [hunk 8]
```diff
@@ -510,14 +527,16 @@ def build_dictionary_from_data_table(
...
+        A field separator character. "" has the same effect as "\t". Sets
+        ``detect_format`` to ``False`` if set. Ignored if ``detect_format``
+        is set to ``True``.
```
Comment (folmos-at-orange) [inline khiops/core/api.py:539]: -> ```Ignored if ``detect_format`` is ``True`` ```.

**Comment by tramora** [comment]
Go for it for this PR despite the ambiguities remaining : https://github.com/KhiopsML/khiops-python/pull/234#issuecomment-2411385695

However the issue https://github.com/KhiopsML/khiops/issues/345 on the khiops core side is not fixed yet (in `10.2.3-b.5`) so the final state remains the same.

**Review by tramora — APPROVED** [review]

**Review by folmos-at-orange — CHANGES_REQUESTED** [review]

File: khiops/core/api.py (hunk 7 above)
Comment (popescu-v) [reply khiops/core/api.py:536]: Ok & Done.

File: khiops/core/api.py (hunk 8 above)
Comment (popescu-v) [reply khiops/core/api.py:539]: Ok & Done.
